### PR TITLE
Send to clients only changed node metadata instad of whole mapblock

### PR DESCRIPTION
**Pull request by RealBadAngel**
thx to sending single meta packet is very small, circa 200-300 bytes for a chest, and it doesnt cause mapblock mesh updates (not only origin block mesh was updated but also neighbours).
Also sending whole block was causing each meta for that block being deleted and restored.
Effect on VE server, known to be laggy as hell and with usually slidshow (~10 fps as on my box):
screenshot_20150909_191925

### CODE REVIEW

File: src/network/networkprotocol.h [hunk 1]
```diff
@@ -620,6 +622,12 @@ enum ToClientCommand
 		u32 id
 	*/
 
+	TOCLIENT_NODEMETA_CHANGED = 0x54,
+	/*
+		u16 command
```
Comment (est31) [inline src/network/networkprotocol.h]: nrz wants to not document the command field, as it exists in any packet (there is no extra information).

File: src/network/serverpackethandler.cpp [hunk 2]
```diff
@@ -925,8 +925,8 @@ void Server::handleCommand_InventoryAction(NetworkPacket* pkt)
 
 		ma->from_inv.applyCurrentPlayer(player->getName());
 		ma->to_inv.applyCurrentPlayer(player->getName());
-
-		setInventoryModified(ma->from_inv, false);
+		if (ma->from_inv != ma->to_inv)
```
Comment (est31) [inline src/network/serverpackethandler.cpp]: Wow, we actually [have implemented](https://github.com/minetest/minetest/blob/7bbb9b066a8cc079512ddd3e6b32475309f49fca/src/inventorymanager.h#L69) this comparator.

File: src/server.cpp [hunk 3]
```diff
@@ -2114,6 +2109,58 @@ void Server::sendAddNode(v3s16 p, MapNode n, u16 ignore_id,
 	}
 }
 
+void Server::sendMetadataChanged(v3s16 p, std::vector<u16> *far_players,
+	float far_d_nodes)
+{
+	v3s16 blockpos = getNodeBlockPos(p);
+	float maxd = far_d_nodes*BS;
```
Comment (est31) [inline src/server.cpp]: space

File: src/server.cpp [hunk 4]
```diff
@@ -2114,6 +2109,58 @@ void Server::sendAddNode(v3s16 p, MapNode n, u16 ignore_id,
...
+	v3f p_f = intToFloat(p, BS);
+	MapBlock *block = m_env->getMap().getBlockNoCreateNoEx(blockpos);
+	if(block)
```
Comment (est31) [inline src/server.cpp]: space

File: src/server.cpp [hunk 5]
```diff
@@ -2114,6 +2109,58 @@ void Server::sendAddNode(v3s16 p, MapNode n, u16 ignore_id,
...
+		block->raiseModified(MOD_STATE_WRITE_NEEDED,
+			MOD_REASON_REPORT_META_CHANGE);
+
+	std::vector<u16> clients = m_clients.getClientIDs();
+	for(std::vector<u16>::iterator i = clients.begin();
```
Comment (est31) [inline src/server.cpp]: here as well, and for the if's following.

File: src/script/lua_api/l_nodemeta.cpp [hunk 6]
```diff
@@ -56,17 +56,10 @@ void NodeMetaRef::reportMetadataChange(NodeMetaRef *ref)
 {
 	// NOTE: This same code is in rollback_interface.cpp
 	// Inform other things that the metadata has changed
-	v3s16 blockpos = getNodeBlockPos(ref->m_p);
 	MapEditEvent event;
 	event.type = MEET_BLOCK_NODE_METADATA_CHANGED;
-	event.p = blockpos;
+	event.p =ref->m_p;
```
Comment (est31) [inline src/script/lua_api/l_nodemeta.cpp]: space here as well

File: src/script/lua_api/l_nodemeta.cpp [hunk 7]
```diff
@@ -56,17 +56,10 @@ void NodeMetaRef::reportMetadataChange(NodeMetaRef *ref)
...
 	ref->m_env->getMap().dispatchEvent(&event);
-	// Set the block to be saved
```
Comment (est31) [inline src/script/lua_api/l_nodemeta.cpp]: why did you remove this? It is needed, so that we store the block.

**Comment by est31** [comment]
You'll have to adjust how the p param MEET_BLOCK_NODE_METADATA_CHANGED is handled (single position instead of blockpos) [everywhere it is used](https://github.com/minetest/minetest/search?utf8=%E2%9C%93&q=MEET_BLOCK_NODE_METADATA_CHANGED).

**Comment by est31** [comment]
Most of the PR looks good, except of things I've already pointed out above, but generally we should be very cautious with applying this. @kwolekr can you review this PR?

File: src/server.cpp [hunk 8]
```diff
@@ -1327,13 +1328,7 @@ void Server::setInventoryModified(const InventoryLocation &loc, bool playerSend)
 		break;
 	case InventoryLocation::NODEMETA:
 	{
-		v3s16 blockpos = getNodeBlockPos(loc.p);
-
-		MapBlock *block = m_env->getMap().getBlockNoCreateNoEx(blockpos);
-		if(block)
-			block->raiseModified(MOD_STATE_WRITE_NEEDED);
```
Comment (est31) [inline src/server.cpp]: same here, we have to take care of storage of the modified block in the map database. We can spare sending it, thats fine, but we shouldn't avoid storing them into the map.

File: src/network/networkprotocol.h (hunk 1 above)
Comment (nerzhul) [inline src/network/networkprotocol.h]: @est31 this is exact, command field is not useful information there

**Comment by RealBadAngel** [comment]
IMHO I dont have to comment using p parameter as single node coords for this event, it means it for each other case already, MEET_BLOCK_NODE_METADATA_CHANGED was an exception to that rule so it was documented.
I have removed raising block as modified from l_nodemeta.cpp and rollback_interface.cpp because its done in the event handler, no need to mark the block three times.
As for the command in packet description: all others have command mentioned in it, why i should format it other way?

**Comment by est31** [comment]
Okay, i see you call `raiseModified`, sorry. And I agree now that it doesn't has to be documented that p indicates the position, I had removed my comment.

**Comment by est31** [comment]
> As for the command in packet description: all others have command mentioned in it, why i should format it other way? 

Not all others, all new and modified packets have the command not mentioned.

**Comment by est31** [comment]
> As for the command in packet description: all others have command mentioned in it, why i should format it other way? 

Now its changed: 64a5eec1bd92a352995ac510b69e6051421ee774

**Comment by RealBadAngel** [comment]
Ive removed "command" from packet desc

**Comment by VanessaE** [comment]
Provides a small improvement for me - a few fps better than before anyway.  Motion seems smoother, less jerky.

File: src/network/clientpackethandler.cpp [hunk 9]
```diff
@@ -277,6 +277,29 @@ void Client::handleCommand_AddNode(NetworkPacket* pkt)
...
+	v3s16 p;
+	*pkt >> p;
+	std::string datastring = pkt->readLongString();
+	std::istringstream is(datastring, std::ios::binary);
+	std::ostringstream oss(std::ios::binary);
+	decompressZlib(is, oss);
+	std::istringstream iss(oss.str(), std::ios::binary);
+	meta->deSerialize(iss);
+	try {
+		m_env.getMap().setNodeMetadata(p, meta);
```
Comment (nerzhul) [inline src/network/clientpackethandler.cpp]: if this return false you MUST delete meta pointer

File: src/server.cpp [hunk 10]
```diff
@@ -2114,6 +2109,58 @@ void Server::sendAddNode(v3s16 p, MapNode n, u16 ignore_id,
...
+	MapBlock *block = m_env->getMap().getBlockNoCreateNoEx(blockpos);
+	if (block)
+		block->raiseModified(MOD_STATE_WRITE_NEEDED,
+			MOD_REASON_REPORT_META_CHANGE);
+
+	NetworkPacket pkt(TOCLIENT_NODEMETA_CHANGED, 0);
+	std::vector<u16> clients = m_clients.getClientIDs();
+
+	if (clients.size() > 0) {
+		NodeMetadata *meta = m_env->getMap().getNodeMetadata(p);
```
Comment (nerzhul) [inline src/server.cpp]: meta pointer could be null, please handle it

**Comment by RealBadAngel** [comment]
@VanessaE and for info to all others: she really has problem with drivers, with her decent HW setup she can barely hit 40 fps....

**Comment by cheapie** [comment]
@RealBadAngel Then why do her and I both get *way* better frame rates with other games, even ones with much fancier graphics?

**Comment by HybridDog** [comment]
minetest uses irrlicht although the world is not a fixed world

**Comment by est31** [comment]
Just wanting to point out that the bug with this PR that you get all metadata changes on the whole server, not just the ones you are close to still remains unfixed.

File: src/server.cpp (hunk 8 above)
Comment (SmallJoker) [inline src/server.cpp]: Done in `Server::sendMetadataChanged`

Comment (est31) [inline src/server.cpp]: yea

**Comment by Fixer-007** [comment]
Anybody has EXE for windows with this feature enabled for testing? Thanks

File: src/network/clientpackethandler.cpp [hunk 11]
```diff
@@ -277,6 +277,29 @@ void Client::handleCommand_AddNode(NetworkPacket* pkt)
...
 	addNode(p, n, remove_metadata);
 }
+
+void Client::handleCommand_NodemetaChanged(NetworkPacket *pkt)
+{
+	if (pkt->getSize() < 6)
+		return;
+
+
+	NodeMetadata *meta = new NodeMetadata(this);
```
Comment (SmallJoker) [inline src/network/clientpackethandler.cpp]: `Unable to convert char* to IItemDefManager*`, changed to `m_itemdef`

**Comment by SmallJoker** [comment]
@Fixer-007 It required some manual changes to merge it with upstream.
However, here's your [Win32 binary, *.7z archive](https://www.dropbox.com/s/zqdq346cxccjitj/minetest_151231_P3166.7z)

**Comment by Fixer-007** [comment]
Thank you very much!

**Comment by RealBadAngel** [comment]
continued in  #3775
